metricsaggregator/src/main/java/io/dockstore/metricsaggregator/MetricsAggregatorS3Client.java

```diff
@@ -0,0 +1,149 @@
+import io.dockstore.openapi.client.model.Execution;
+import io.dockstore.webservice.core.Partner;
+import io.dockstore.webservice.core.metrics.MetricsData;
+import io.dockstore.webservice.core.metrics.MetricsDataS3Client;
+import io.dockstore.webservice.helpers.S3ClientHelper;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.CommonPrefix;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
+
+public class MetricsAggregatorS3Client {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MetricsAggregatorS3Client.class);
+    private static final Gson GSON = new Gson();
+
+    private final String bucketName;
+
+    private final S3Client s3Client;
+    private final MetricsDataS3Client metricsDataS3Client;
+
+    public MetricsAggregatorS3Client(String bucketName) {
+        this.bucketName = bucketName;
+        this.s3Client = S3ClientHelper.createS3Client();
+        this.metricsDataS3Client = new MetricsDataS3Client(bucketName, this.s3Client);
+    }
+
+    public MetricsAggregatorS3Client(String bucketName, String s3EndpointOverride) throws URISyntaxException {
+        this.bucketName = bucketName;
+        this.s3Client = S3ClientHelper.createS3Client(s3EndpointOverride);
+        this.metricsDataS3Client = new MetricsDataS3Client(bucketName, s3EndpointOverride);
+    }
+
+    public void aggregateMetrics(ExtendedGa4GhApi extendedGa4GhApi) {
+        List<String> metricsDirectories = getDirectories();
+        // Each directory contains metrics for a specific tool version and platform
+        for (String directory : metricsDirectories) {
+            String toolId = S3ClientHelper.getToolId(directory); // Check if we should just give the full key
+            String versionName = S3ClientHelper.getVersionName(directory);
+            String platform = S3ClientHelper.getPlatform(directory);
+            List<MetricsData> metricsDataList = metricsDataS3Client.getMetricsData(toolId, versionName, Partner.valueOf(platform));
+            List<Execution> executions = metricsDataList.stream().map(metricsData -> {
+                try {
+                    String fileContent = metricsDataS3Client.getMetricsDataFileContent(metricsData.toolId(), metricsData.toolVersionName(),
+                            platform, metricsData.fileName());
+                    return List.of(GSON.fromJson(fileContent, Execution[].class));
+                } catch (IOException e) {
+                    LOG.error("Error aggregating metrics: Unable to get all executions", e);
+                    throw new RuntimeException("Error aggregating metrics: Unable to get all executions");
```

Comment:
Just making sure, do you want to stop all processing because one failed?

Reply:
No, changed it so it logs an error and continues processing for other S3 directories